core/common/src/grpc_interceptor.rs

```diff
@@ -0,0 +1,238 @@
+        Self { interceptor_factory }
+    }
+}
+
+impl<S> Layer<S> for GrpcInterceptorLayer {
+    type Service = GrpcInterceptorService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        GrpcInterceptorService { service, interceptor_factory: self.interceptor_factory }
+    }
+}
+
+/// The tower service that hosts a gRPC Interceptor.
+#[derive(Clone)]
+pub struct GrpcInterceptorService<S> {
+    service: S,
+    interceptor_factory: GrpcInterceptorFactory,
+}
+
+impl<S> GrpcInterceptorService<S> {
+    /// Retrieve the gRPC service name and method name from a URI.
+    ///
+    /// * `uri` - The uri used for the gRPC call.
+    fn retrieve_grpc_names_from_uri(uri: &Uri) -> (String, String) {
+        let uri_string = uri.to_string();
+        let uri_parts: Vec<&str> = uri_string.split('/').collect();
+        let mut service_name = String::new();
+        let mut method_name = String::new();
+        if uri_parts.len() == 5 {
+            method_name = uri_parts[4].to_string();
+            let qualified_service_name = uri_parts[3].to_string();
+            let qualified_service_name_parts: Vec<&str> =
+                qualified_service_name.split('.').collect();
+            if qualified_service_name_parts.len() == 2 {
+                service_name = qualified_service_name_parts[1].to_string();
+            }
+        }
+        (service_name, method_name)
+    }
+}
+
+impl<S> Service<http::request::Request<tonic::transport::Body>> for GrpcInterceptorService<S>
+where
+    S: Service<
+            http::request::Request<tonic::transport::Body>,
+            Response = http::response::Response<tonic::body::BoxBody>,
+            Error = Box<dyn std::error::Error + Sync + Send>,
+        > + Send,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    /// Implementation of tower's Service trait's poll_ready method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    /// Implementation of tower's Service trait's call method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn call(
+        &mut self,
+        mut request: http::request::Request<tonic::transport::Body>,
+    ) -> Self::Future {
+        let interceptor = (self.interceptor_factory)();
+
+        let (service_name, method_name) = Self::retrieve_grpc_names_from_uri(request.uri());
+        let is_applicable = interceptor.is_applicable(&service_name, &method_name);
+
+        if is_applicable && interceptor.must_handle_request() {
+            let (parts, body) = request.into_parts();
+            let mut body_bytes: Bytes =
+                match futures::executor::block_on(hyper::body::to_bytes(body)) {
```

Comment:
this is a job for `map_err` and `?`. I think you can also omit the `: Bytes` type declaration above:

```rust
let mut body_bytes = futures::executor::block_on(hyper::body::to_bytes(body))
    .map_err(|err| {
        Box::pin(async move {
            Err(Box::new(err) as _) // or if "as _" doesn't work, then what you've got already
        })
    })?;
```

Reply:
Here is the error that I received when I tried your code:

    | |       ^ cannot use the `?` operator in a method that returns `Pin<Box<dyn futures_core::Future<Output = Result<http::Response<UnsyncBoxBody<bytes::Bytes, Status>>, Box<(dyn std::error::Error + Sync + std::marker::Send + 'static)>>> + std::marker::Send>>`

Reply:
ok, I think this suggestion won't work unless there's a `Service` implementation that exposes an async version of this trait. I'd say leave it as-is